src/engine/server_iv.c

```diff
@@ -627,8 +627,10 @@ ivc_on_put(crt_iv_namespace_t ivns, d_sg_list_t *iv_value, void *priv)
 	int			 rc;
 
 	rc = iv_ns_lookup_by_ivns(ivns, &ns);
-	if (rc != 0)
+	if (rc != 0) {
+		ds_iv_ns_put(ns); /* balance ivc_on_get */
 		return rc;
+	}
```

Comment:
That does not look correct to me. The code in iv_ns_lookup_by_ivns does this:
```
 186 iv_ns_lookup_by_ivns(crt_iv_namespace_t ivns, struct ds_iv_ns **p_ns)
 187 {
 188         struct ds_iv_ns *ns;
 189
 190         *p_ns = NULL;
 191         d_list_for_each_entry(ns, &ds_iv_ns_list, iv_ns_link) {
 192                 if (ns->iv_ns != ivns)
 193                         continue;
 194
 195                 if (ns->iv_stop) {
 196                         D_DEBUG(DB_MD, DF_UUID" stopping\n",
 197                                 DP_UUID(ns->iv_pool_uuid));
 198                         return -DER_SHUTDOWN;
 199                 }
 200                 ds_iv_ns_get(ns);
 201                 *p_ns = ns;
 202                 return 0;
 203         }
 204         return -DER_NONEXIST;
```
 
From this it looks like it will only do ds_iv_ns_get() if it is successful on line 200

Reply:
if it comes to ivc_on_put(), then it means ivc_on_get() succeeds, so this put is for balancing the get in ivc_on_get(), instead of iv_ns_lookup_by_ivns().  See there is two ds_iv_ns_put(ns) at the end of this function.

Reply:
No, they put twice at the end of this function. One is for this get (iv_ns_lookup_by_ivns), and the second should be always.

Reply:
Ah, I missed setting the *p_ns for SHUTDOWN case, which will cause segfault